src/tracer.js

```diff
@@ -0,0 +1,258 @@
+    *        be unspecified.
+    * @param {array} [fields.references] - an array of Reference instances,
+    *        each pointing to a causal parent SpanContext. If specified,
+    *        `fields.childOf` must be unspecified.
+    * @param {object} [fields.tags] - set of key-value pairs which will be set
+    *        as tags on the newly created Span. Ownership of the object is
+    *        passed to the created span for efficiency reasons (the caller
+    *        should not modify this object after calling startSpan).
+    * @param {number} [fields.startTime] - a manually specified start time for
+    *        the created Span object. The time should be specified in
+    *        milliseconds as Unix timestamp. Decimal value are supported
+    *        to represent time values with sub-millisecond accuracy.
+    * @return {Span} - a new Span object.
+    **/
+    startSpan(fields: spanFields): Span {
+        let tags = fields.tags || {}
+        let references = fields.references || [];
+        let startTime = fields.startTime;
+        if (!startTime) {
+            startTime = Utils.getTimestampMicros();
+        }
+
+        let parent: ?SpanContext;
+        for (let i = 0; i < references.length; i++) {
+            let ref: Reference = references[i];
+            if (ref.type() === opentracing.REFERENCE_CHILD_OF) {
+                parent = ref.referencedContext();
+                break;
+            } else {
+                // todo(oibe) support other types of span references
+            }
+        }
+
+        let rpcServer = false;
+        let spanKindValue = tags[opentracing_tags.SPAN_KIND];
+        if (spanKindValue) {
+            rpcServer = (spanKindValue === opentracing_tags.SPAN_KIND_RPC_SERVER);
+        }
+
+
+        // $FlowIgnore - I just want a span context up front.
+        let ctx: SpanContext = new SpanContext();
+        if (!parent) {
+            let randomId = Utils.getRandom64();
+            let flags = 0;
+            if (this._sampler.isSampled()) {
+                flags |= constants.SAMPLED_MASK;
+            }
+
+            ctx.traceId = randomId;
+            ctx.spanId = randomId;
+            ctx.parentId = null;
+            ctx.flags = flags;
+        } else {
+            ctx.traceId = parent.traceId;
+            if (rpcServer) {
+                // Support Zipkin's one-span-per-RPC model
+                ctx.spanId = parent.spanId;
+                ctx.parentId = parent.parentId;
+            } else {
+                let randomId = Utils.getRandom64();
+                ctx.spanId = randomId;
+                ctx.parentId = parent.spanId;
+            }
+            ctx.flags = parent.flags;
+
+            // copy baggage items
+            for (let key in parent.baggage) {
+                if (parent.baggage.hasOwnProperty(key)) {
+                    let value = parent.baggage[key];
+                    ctx.setBaggageItem(key, value);
```

Comment:
if this was python, all we really need is `ctx.baggage = dict(parent.baggage)`. I don't know if JS has a dict copy, but we certainly don't need to go through setBaggageItem method, because we already have a normalized dictionary from parent.

Once we (do)/land the PR for "immutable" SpanContext, this can simply become `ctx.baggage = parent.baggage`, because span.setBaggage() should be "copy on write", which happens a lot less often than creating new spans from parent spans.